File: ES01-Jakarta/src/main/java/it/distributedsystems/model/ejb/EJB3ProducerDAO.java
package it.distributedsystems.model.ejb;

//import it.distributedsystems.model.logging.OperationLogger;
import it.distributedsystems.model.dao.Customer;
import it.distributedsystems.model.dao.CustomerDAO;
import it.distributedsystems.model.dao.Producer;
import it.distributedsystems.model.dao.ProducerDAO;
import it.distributedsystems.util.GeneralUtils;
import jakarta.ejb.Local;
import jakarta.ejb.Stateless;
import jakarta.ejb.TransactionAttribute;
import jakarta.ejb.TransactionAttributeType;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;

import java.util.List;

@Stateless
@Local(ProducerDAO.class)
//@Remote(CustomerDAO.class) //-> TODO: serve nella versione clustering???
public class EJB3ProducerDAO implements ProducerDAO {

    @PersistenceContext(unitName = "distributed-systems-demo")
    EntityManager em;

    @Override
//    @Interceptors(OperationLogger.class)
    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    public int insertProducer(Producer producer) {
        em.persist(producer);
        return producer.getId();
    }

    @Override
    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    public int removeProducerByName(String name) {

        Producer producer;
        if(name != null && !name.equals("")) {
            producer = (Producer) em.createQuery("FROM Producer p WHERE p.name = :producerName")
                    .setParameter("producerName", name)
                    .getSingleResult();
            em.remove(producer);
            return producer.getId();
        } else
            return 0;
    }

    @Override
    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    public int removeProducerById(int id) {
        Producer producer = em.find(Producer.class, id);
        if (producer!=null){
            em.remove(producer);
            return id;
        }
        else
            return 0;
    }

    @Override
    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    public Producer findProducerByName(String name) {
        if(name != null && !name.equals("")) {
            return (Producer) em.createQuery("FROM Producer p where p.name = :producerName").
                    setParameter("producerName", name).getSingleResult();
        } else
            return null;
    }

    @Override
    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    public Producer findProducerById(int id) {
        return em.find(Producer.class, id);
    }

    @Override
    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    public List<Producer> getAllProducers() {
        return GeneralUtils.withNullAsEmptyList(em.createQuery("FROM Producer").getResultList());
    }
}

File: Ex1/src/main/kotlin/it/distributedsystems/model/ejb/EJB3ProducerDAO.kt
package it.distributedsystems.model.ejb

import it.distributedsystems.model.dao.CustomerDAO
import it.distributedsystems.model.dao.Producer
import it.distributedsystems.model.dao.ProducerDAO
import javax.ejb.Local
import javax.ejb.Stateless
import javax.ejb.TransactionAttribute
import javax.ejb.TransactionAttributeType
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

//import it.distributedsystems.model.logging.OperationLogger;
@Stateless
@Local(CustomerDAO::class) //@Remote(CustomerDAO.class) //-> TODO: serve nella versione clustering???
class EJB3ProducerDAO : ProducerDAO {
    @PersistenceContext(unitName = "distributed-systems-demo")
    var em: EntityManager? = null

    //    @Interceptors(OperationLogger.class)
    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    override fun insertProducer(producer: Producer): Int {
        em!!.persist(producer)
        return producer.id
    }

    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    override fun removeProducerByName(name: String): Int {
        val producer: Producer
        return if (name != null && name != "") {
            producer = em!!.createQuery("FROM Producer p WHERE p.name = :producerName")
                .setParameter("producerName", name)
                .singleResult as Producer
            em!!.remove(producer)
            producer.id
        } else 0
    }

    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    override fun removeProducerById(id: Int): Int {
        val producer = em!!.find(
            Producer::class.java, id
        )
        return if (producer != null) {
            em!!.remove(producer)
            id
        } else 0
    }

    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    override fun findProducerByName(name: String): Producer? {
        return if (name != null && name != "") {
            em!!.createQuery("FROM Producer p where p.name = :producerName")
                .setParameter("producerName", name).singleResult as Producer
        } else null
    }

    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    override fun findProducerById(id: Int): Producer? {
        return em!!.find(Producer::class.java, id)
    }

    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    override fun getAllProducers(): List<Producer> {
        return em!!.createQuery("FROM Producer").resultList as List<Producer>
    }
}